File: src/brs/web/api/http/handler/GetTrades.java
package brs.web.api.http.handler;

import brs.Account;
import brs.Asset;
import brs.SignumException;
import brs.Trade;
import brs.assetexchange.AssetExchange;
import brs.web.api.http.common.LegacyDocTag;
import brs.web.api.http.common.ParameterParser;
import brs.web.api.http.common.Parameters;
import brs.services.ParameterService;
import brs.util.CollectionWithIndex;
import brs.util.Convert;

import brs.web.api.http.ApiServlet;
import brs.web.api.http.common.JSONData;
import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;

import javax.servlet.http.HttpServletRequest;

import static brs.web.api.http.common.Parameters.*;
import static brs.web.api.http.common.ResultFields.NEXT_INDEX_RESPONSE;
import static brs.web.api.http.common.ResultFields.TRADES_RESPONSE;

public final class GetTrades extends ApiServlet.JsonRequestHandler {

  private final ParameterService parameterService;
  private final AssetExchange assetExchange;

  public GetTrades(ParameterService parameterService, AssetExchange assetExchange) {
    super(new LegacyDocTag[] {LegacyDocTag.AE}, ASSET_PARAMETER, ACCOUNT_PARAMETER, FIRST_INDEX_PARAMETER, LAST_INDEX_PARAMETER, INCLUDE_ASSET_INFO_PARAMETER);
    this.parameterService = parameterService;
    this.assetExchange = assetExchange;
  }

  @Override
  protected
  JsonElement processRequest(HttpServletRequest req) throws SignumException {

    String assetId = Convert.emptyToNull(req.getParameter(ASSET_PARAMETER));
    String accountId = Convert.emptyToNull(req.getParameter(ACCOUNT_PARAMETER));

    int firstIndex = ParameterParser.getFirstIndex(req);
    int lastIndex = ParameterParser.getLastIndex(req);
    boolean includeAssetInfo = !Parameters.isFalse(req.getParameter(INCLUDE_ASSET_INFO_PARAMETER));

    JsonObject response = new JsonObject();
    JsonArray tradesData = new JsonArray();
    CollectionWithIndex<Trade> trades;
    Asset asset = null;
    if (accountId == null) {
      asset = parameterService.getAsset(req);
      trades = assetExchange.getTrades(asset.getId(), firstIndex, lastIndex);
    } else if (assetId == null) {
      Account account = parameterService.getAccount(req);
      trades = assetExchange.getAccountTrades(account.getId(), firstIndex, lastIndex);
    } else {
      asset = parameterService.getAsset(req);
      Account account = parameterService.getAccount(req);
      trades = assetExchange.getAccountAssetTrades(account.getId(), asset.getId(), firstIndex, lastIndex);
    }
    if(!includeAssetInfo) {
      asset = null;
    }
    for (Trade trade : trades) {
      if(includeAssetInfo && (asset == null || asset.getId() != trade.getAssetId())) {
        asset = assetExchange.getAsset(trade.getAssetId());
      }
      tradesData.add(JSONData.trade(trade, asset));
    }
    response.add(TRADES_RESPONSE, tradesData);
    if(trades.hasNextIndex()) {
      response.addProperty(NEXT_INDEX_RESPONSE, trades.nextIndex());
    }

    return response;
  }
}

File: src/main/java/brs/http/GetTrades.kt
package brs.http

import brs.Trade
import brs.assetexchange.AssetExchange
import brs.http.common.Parameters.ACCOUNT_PARAMETER
import brs.http.common.Parameters.ASSET_PARAMETER
import brs.http.common.Parameters.FIRST_INDEX_PARAMETER
import brs.http.common.Parameters.INCLUDE_ASSET_INFO_PARAMETER
import brs.http.common.Parameters.LAST_INDEX_PARAMETER
import brs.http.common.Parameters.isFalse
import brs.http.common.ResultFields.TRADES_RESPONSE
import brs.services.ParameterService
import brs.util.Convert
import com.google.gson.JsonArray
import com.google.gson.JsonElement
import com.google.gson.JsonObject
import javax.servlet.http.HttpServletRequest

internal class GetTrades internal constructor(private val parameterService: ParameterService, private val assetExchange: AssetExchange) : APIServlet.JsonRequestHandler(arrayOf(APITag.AE), ASSET_PARAMETER, ACCOUNT_PARAMETER, FIRST_INDEX_PARAMETER, LAST_INDEX_PARAMETER, INCLUDE_ASSET_INFO_PARAMETER) {

    override suspend fun processRequest(request: HttpServletRequest): JsonElement {

        val assetId = Convert.emptyToNull(request.getParameter(ASSET_PARAMETER))
        val accountId = Convert.emptyToNull(request.getParameter(ACCOUNT_PARAMETER))

        val firstIndex = ParameterParser.getFirstIndex(request)
        val lastIndex = ParameterParser.getLastIndex(request)
        val includeAssetInfo = !isFalse(request.getParameter(INCLUDE_ASSET_INFO_PARAMETER))

        val response = JsonObject()
        val tradesData = JsonArray()
        val trades: Collection<Trade>
        trades = if (accountId == null) {
            val asset = parameterService.getAsset(request)
            assetExchange.getTrades(asset.id, firstIndex, lastIndex)
        } else if (assetId == null) {
            val account = parameterService.getAccount(request) ?: return JSONResponses.INCORRECT_ACCOUNT
            assetExchange.getAccountTrades(account.id, firstIndex, lastIndex)
        } else {
            val asset = parameterService.getAsset(request)
            val account = parameterService.getAccount(request) ?: return JSONResponses.INCORRECT_ACCOUNT
            assetExchange.getAccountAssetTrades(account.id, asset.id, firstIndex, lastIndex)
        }
        for (trade in trades) {
            val asset = if (includeAssetInfo) assetExchange.getAsset(trade.assetId) else null
            tradesData.add(JSONData.trade(trade, asset))
        }
        response.add(TRADES_RESPONSE, tradesData)

        return response
    }
}
